src/dbnode/storage/index/read_through_segment.go

```diff
@@ -0,0 +1,183 @@
+// ReadThroughSegment.
+type ReadThroughSegmentOptions struct {
+	// Whether the postings list for regexp queries should be cached.
+	CacheRegexp bool
+	// Whether the postings list for term queries should be cached.
+	CacheTerms bool
+}
+
+// NewReadThroughSegment creates a new read through segment.
+func NewReadThroughSegment(
+	seg segment.Segment,
+	cache *PostingsListCache,
+	opts ReadThroughSegmentOptions,
+) segment.Segment {
+	return &ReadThroughSegment{
+		Segment: seg,
+
+		opts:              opts,
+		uuid:              uuid.NewUUID(),
+		postingsListCache: cache,
+	}
+}
+
+// Reader returns a read through reader for the read through segment.
+func (r *ReadThroughSegment) Reader() (index.Reader, error) {
+	r.RLock()
+	defer r.RUnlock()
+	if r.closed {
+		return nil, errCantGetReaderFromClosedSegment
+	}
+
+	reader, err := r.Segment.Reader()
+	if err != nil {
+		return nil, err
+	}
+	return newReadThroughSegmentReader(
+		reader, r.uuid, r.postingsListCache, r.opts), nil
+}
+
+// Close purges all entries in the cache associated with this segment,
+// and then closes the underlying segment.
+func (r *ReadThroughSegment) Close() error {
+	r.Lock()
+	defer r.Unlock()
+	if r.closed {
+		return errCantCloseClosedSegment
+	}
+
+	r.closed = true
+
+	if r.postingsListCache != nil {
+		// Purge segments from the cache before closing the segment to avoid
+		// temporarily having postings lists in the cache whose underlying
+		// bytes are no longer mmap'd.
+		r.postingsListCache.PurgeSegment(r.uuid)
+	}
+	return r.Segment.Close()
+}
+
+type readThroughSegmentReader struct {
+	index.Reader
+
+	opts              ReadThroughSegmentOptions
+	uuid              uuid.UUID
+	postingsListCache *PostingsListCache
+}
+
+func newReadThroughSegmentReader(
+	reader index.Reader,
+	uuid uuid.UUID,
+	cache *PostingsListCache,
+	opts ReadThroughSegmentOptions,
+) index.Reader {
+	return &readThroughSegmentReader{
+		Reader: reader,
+
+		opts:              opts,
+		uuid:              uuid,
+		postingsListCache: cache,
+	}
+}
+
+// MatchRegexp returns a cached posting list or queries the underlying
+// segment if their is a cache miss.
+func (s *readThroughSegmentReader) MatchRegexp(
+	field []byte,
+	c index.CompiledRegex,
+) (postings.List, error) {
+	if s.postingsListCache == nil || !s.opts.CacheRegexp {
+		return s.Reader.MatchRegexp(field, c)
+	}
+
+	pattern := c.FSTSyntax.String()
+	pl, ok := s.postingsListCache.GetRegexp(s.uuid, pattern)
+	if ok {
+		return pl, nil
+	}
+
+	pl, err := s.Reader.MatchRegexp(field, c)
+	if err == nil {
+		s.postingsListCache.PutRegexp(s.uuid, pattern, pl)
+	}
+	return pl, err
+}
+
+// MatchTerm returns a cached posting list or queries the underlying
+// segment if their is a cache miss.
+func (s *readThroughSegmentReader) MatchTerm(
+	field []byte, term []byte,
+) (postings.List, error) {
+	if s.postingsListCache == nil || !s.opts.CacheTerms {
+		return s.Reader.MatchTerm(field, term)
+	}
+
+	termString := string(term)
```

Comment:
can you avoid these allocs?

Reply:
we could, but not without doing something complicated or "unsafe" right now. I'm not worried about it, compared to the 50GiB in postings lists we're still allocating every few minutes this is nothing and I'd rather avoid more complexity here and focus on solving the other performance issues

Reply:
👍 sg

Reply:
maybe add TODO as a reminder for the future?

Reply:
Sure left a TODO